sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -283,8 +283,56 @@ pi_result piDeviceGetInfo(pi_device device, pi_device_info paramName,
     // sycl/doc/extensions/supported/sycl_ext_intel_device_info.md
   case PI_DEVICE_INFO_UUID:
   case PI_DEVICE_INFO_ATOMIC_MEMORY_ORDER_CAPABILITIES:
-  case PI_DEVICE_INFO_ATOMIC_MEMORY_SCOPE_CAPABILITIES:
     return PI_ERROR_INVALID_VALUE;
+  case PI_DEVICE_INFO_ATOMIC_MEMORY_SCOPE_CAPABILITIES: {
+    // Initialize result to minimum mandated capabilities according to
+    // SYCL2020 4.6.3.2
+    pi_memory_scope_capabilities result = PI_MEMORY_SCOPE_WORK_ITEM |
+                                          PI_MEMORY_SCOPE_SUB_GROUP |
+                                          PI_MEMORY_SCOPE_WORK_GROUP;
+
+    OCLV::OpenCLVersion devVer;
+
+    cl_device_id deviceID = cast<cl_device_id>(device);
+    cl_int ret_err = getDeviceVersion(deviceID, devVer);
+    if (ret_err != CL_SUCCESS)
+      return static_cast<pi_result>(ret_err);
+
+    pi_device_atomic_capabilities devCapabilities = 0;
+    if (devVer >= OCLV::V3_0) {
+      ret_err = clGetDeviceInfo(deviceID, PI_DEVICE_ATOMIC_MEMORY_CAPABILITIES,
+                                sizeof(pi_device_atomic_capabilities),
+                                &devCapabilities, nullptr);
+      if (ret_err != CL_SUCCESS)
+        return static_cast<pi_result>(ret_err);
+      assert(devCapabilities && PI_DEVICE_ATOMIC_SCOPE_WORK_GROUP &&
+             "Violates minimum mandated guarantee");
+
+      // Because scopes are hierarchical, wider scopes support all narrower
+      // scopes. SUB_GROUP and WORK_ITEM were already included in the
+      // initialization, since WORK_GROUP is mandated minimum capality.
+      if (devCapabilities && PI_DEVICE_ATOMIC_SCOPE_DEVICE) {
+        result |= PI_MEMORY_SCOPE_DEVICE;
+      }
```

Comment:
Oh, I suppose that's the answer to my question above about default value. Do you have a pointer to spec section which describes the hierarchical nature of those scopes?

Reply:
This comes from OpenCL Spec: https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_DEVICE_ATOMIC_MEMORY_CAPABILITIES

However, you can see there that `WORK_ITEM` is actually an exception in OpenCL. I had an offline discussion with @gmlueck, @bashbaug and @Pennycook about this, and the conclusion is that SYCL should always return `WORK_ITEM`. I think there will be an update in the spec to detail this.

Reply:
Got it, thanks! Perhaps it would be better to put that link to OCL spec directly into sources next to the comment

Reply:
I thought about that, but the OCL spec explicitly says `WORK_ITEM` is an exception, so that might be confusing because we are not treating it as an exception. When SYCL spec is updated, we can obviously add a ref to that next to the comment. For now, I'm not sure if referencing OCL spec is helpful or more confusing. What do you think?

Reply:
Technically, this code is within OpenCL plugin and therefore I think it shouldn't be too confusing to leave links to OpenCL spec in there.

Reply:
Discussed this offline. There is a corner case with `WORK_ITEM` scope being defined differently between SYCL and OpenCL, which is a reason for ignoring this OpenCL exception.

I think that it would still make sense to leave an OCL spec link, but we should have a comment about why we ignore that `work-item` exception. If we already have a link to SYCL spec clarification PR/issue, then even better, we can link it there as well.

Reply:
I extended the current comment with further explanation and a link to OCL spec. Please, let me know if this is enough.